File: UnitTestCSharp/TestGeoCoordinate2d.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using QuadKeys;

namespace QuadKeyTests
{
    [TestClass]
    public class TestGeoCoordinate2d
    {
        [TestMethod]
        public void TestDefaultCtor()
        {
            GeoCoordinate2d empty = new GeoCoordinate2d();
            Assert.IsTrue(new Longitude(0.0).EqualTo(empty.longitude()));
            Assert.IsTrue(new Latitude(0.0).EqualTo(empty.latitude()));
        }

        [TestMethod]
        public void TestCtor()
        {
            {
                GeoCoordinate2d coord = new GeoCoordinate2d(new Longitude(0.0),
                    new Latitude(0.0));
                Assert.IsTrue(new Longitude(0.0).EqualTo(coord.longitude()));
                Assert.IsTrue(new Latitude(0.0).EqualTo(coord.latitude()));
            }

            {
                GeoCoordinate2d coord = new GeoCoordinate2d(
                    new Longitude(-180.001), new Latitude(-95.0));
                Assert.IsTrue(new Longitude(-180.0).EqualTo(
                    coord.longitude()));
                Assert.IsTrue(new Latitude(-90.0).EqualTo(coord.latitude()));
            }

            {
                GeoCoordinate2d coord = new GeoCoordinate2d(
                    new Longitude(-180.001), new Latitude(95.0));
                Assert.IsTrue(new Longitude(-180.0).EqualTo(
                    coord.longitude()));
                Assert.IsTrue(new Latitude(90.0).EqualTo(coord.latitude()));
            }

            {
                GeoCoordinate2d coord = new GeoCoordinate2d(
                    new Longitude(180.001), new Latitude(95.0));
                Assert.IsTrue(new Longitude(180.0).EqualTo(coord.longitude()));
                Assert.IsTrue(new Latitude(90.0).EqualTo(coord.latitude()));
            }

            {
                GeoCoordinate2d coord = new GeoCoordinate2d(
                    new Longitude(180.001), new Latitude(-95.0));
                Assert.IsTrue(new Longitude(180.0).EqualTo(coord.longitude()));
                Assert.IsTrue(new Latitude(-90.0).EqualTo(coord.latitude()));
            }
        }

        [TestMethod]
        public void TestNear()
        {
            {
                Longitude lon = new Longitude(0.001);
                Assert.IsTrue(lon.near(new Longitude(0.0009), 1e-3));
            }

            {
                Longitude lon = new Longitude(0.001);
                Assert.IsFalse(lon.near(new Longitude(0.0009), 1e-4));
            }

            {
                Latitude lat = new Latitude(0.001);
                Assert.IsTrue(lat.near(new Latitude(0.0009), 1e-3));
            }

            {
                Latitude lat = new Latitude(0.001);
                Assert.IsFalse(lat.near(new Latitude(0.0009), 1e-4));
            }

            {
                GeoCoordinate2d coord = new GeoCoordinate2d(
                    new Longitude(0.001), new Latitude(0.001));
                Assert.IsTrue(coord.near(new GeoCoordinate2d(
                    new Longitude(0.0009), new Latitude(0.0009)), 1e-3));
            }

            {
                GeoCoordinate2d coord = new GeoCoordinate2d(
                    new Longitude(0.001), new Latitude(0.001));
                Assert.IsFalse(coord.near(new GeoCoordinate2d(
                    new Longitude(0.001), new Latitude(0.0009)), 1e-4));
            }

            {
                GeoCoordinate2d  coord = new GeoCoordinate2d(
                    new Longitude(0.001), new Latitude(0.001));
                Assert.IsFalse(coord.near(new GeoCoordinate2d(
                    new Longitude(0.0009), new Latitude(0.001)), 1e-4));
            }
        }

        [TestMethod]
        public void TestOperatorLessThan()
        {
            GeoCoordinate2d sw = new GeoCoordinate2d(
                new Longitude(-0.1), new Latitude(-0.1));
            GeoCoordinate2d ne = new GeoCoordinate2d(
                new Longitude(+0.1), new Latitude(+0.1));

            Assert.IsTrue(sw.longitude().LessThan(ne.longitude()));
            Assert.IsTrue(sw.latitude().LessThan(ne.latitude()));
        }

        [TestMethod]
        public void TestOperatorLessThanOrEqual()
        {
            {
                GeoCoordinate2d sw = new GeoCoordinate2d(
                    new Longitude(-0.1), new Latitude(-0.1));
                GeoCoordinate2d ne = new GeoCoordinate2d(
                    new Longitude(+0.1), new Latitude(+0.1));

                Assert.IsTrue(sw.longitude().LessThanOrEqual(ne.longitude()));
                Assert.IsTrue(sw.latitude().LessThanOrEqual(ne.latitude()));
            }

            {
                GeoCoordinate2d sw = new GeoCoordinate2d(
                    new Longitude(-0.1), new Latitude(-0.1));
                GeoCoordinate2d ne = new GeoCoordinate2d(
                    new Longitude(-0.1), new Latitude(-0.1));

                Assert.IsTrue(sw.longitude().LessThanOrEqual(ne.longitude()));
                Assert.IsTrue(sw.latitude().LessThanOrEqual(ne.latitude()));
            }
        }

        [TestMethod]
        public void TestOperatorEqual()
        {
            GeoCoordinate2d sw = new GeoCoordinate2d(
                new Longitude(-0.1), new Latitude(-0.1));
            GeoCoordinate2d ne = new GeoCoordinate2d(
                new Longitude(-0.1), new Latitude(-0.1));

            Assert.IsTrue(sw.longitude().EqualTo(ne.longitude()));
            Assert.IsTrue(sw.latitude().EqualTo(ne.latitude()));
        }
    }
}

File: bindings/python/quad_keys/tests/test_geo_coordinate2d.py
"""
Unit Tests for GeoCoordinate2d
"""

from __future__ import print_function

import unittest

from quad_keys.quad_keys import GeoCoordinate2d
from quad_keys.quad_keys import Longitude
from quad_keys.quad_keys import Latitude


class TestGeoCoordinate2d(unittest.TestCase):
    """Class for running unit tests."""

    def test_default_ctor(self):
        """Test default ctor."""
        empty = GeoCoordinate2d()
        self.assertEqual(Longitude(0.0), empty.get_longitude())
        self.assertEqual(Latitude(0.0), empty.get_latitude())

    def test_ctor(self):
        """Test ctor."""
        coord = GeoCoordinate2d(Longitude(0.0), Latitude(0.0))
        self.assertEqual(Longitude(0.0), coord.get_longitude())
        self.assertEqual(Latitude(0.0), coord.get_latitude())

        coord = GeoCoordinate2d(Longitude(-180.001), Latitude(-95.0))
        self.assertEqual(Longitude(-180.0), coord.get_longitude())
        self.assertEqual(Latitude(-90.0), coord.get_latitude())

        coord = GeoCoordinate2d(Longitude(-180.001), Latitude(95.0))
        self.assertEqual(Longitude(-180.0), coord.get_longitude())
        self.assertEqual(Latitude(90.0), coord.get_latitude())

        coord = GeoCoordinate2d(Longitude(180.001), Latitude(95.0))
        self.assertEqual(Longitude(180.0), coord.get_longitude())
        self.assertEqual(Latitude(90.0), coord.get_latitude())

        coord = GeoCoordinate2d(Longitude(180.001), Latitude(-95.0))
        self.assertEqual(Longitude(180.0), coord.get_longitude())
        self.assertEqual(Latitude(-90.0), coord.get_latitude())

    def test_near(self):
        """Test to see if one coord is near another."""
        lon = Longitude(0.001)
        self.assertTrue(lon.near(Longitude(0.0009), 1e-3))

        lon = Longitude(0.001)
        self.assertFalse(lon.near(Longitude(0.0009), 1e-4))

        lat = Latitude(0.001)
        self.assertTrue(lat.near(Latitude(0.0009), 1e-3))

        lat = Latitude(0.001)
        self.assertFalse(lat.near(Latitude(0.0009), 1e-4))

        coord = GeoCoordinate2d(Longitude(0.001), Latitude(0.001))
        self.assertTrue(coord.near(
            GeoCoordinate2d(Longitude(0.0009), Latitude(0.0009)), 1e-3))

        coord = GeoCoordinate2d(Longitude(0.001), Latitude(0.001))
        self.assertFalse(coord.near(
            GeoCoordinate2d(Longitude(0.001), Latitude(0.0009)), 1e-4))

        coord = GeoCoordinate2d(Longitude(0.001), Latitude(0.001))
        self.assertFalse(coord.near(
            GeoCoordinate2d(Longitude(0.0009), Latitude(0.001)), 1e-4))

    def test_operator_less_than(self):
        """Tests less than operator."""
        sw = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        ne = GeoCoordinate2d(Longitude(+0.1), Latitude(+0.1))

        self.assertTrue(sw.get_longitude() < ne.get_longitude())
        self.assertTrue(sw.get_latitude() < ne.get_latitude())

    def test_operator_less_than_or_equal(self):
        """Tests less than or equal operator."""
        sw = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        ne = GeoCoordinate2d(Longitude(+0.1), Latitude(+0.1))
        self.assertTrue(sw.get_longitude() <= ne.get_longitude())
        self.assertTrue(sw.get_latitude() <= ne.get_latitude())

        sw = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        ne = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        self.assertTrue(sw.get_longitude() <= ne.get_longitude())
        self.assertTrue(sw.get_latitude() <= ne.get_latitude())

    def test_operator_equal(self):
        """Tests for equality of GeoCoordinate2d."""
        sw = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        ne = GeoCoordinate2d(Longitude(-0.1), Latitude(-0.1))
        self.assertEqual(sw.get_longitude(), ne.get_longitude())
        self.assertEqual(sw.get_latitude(), ne.get_latitude())
